src/sqlfluff/dialects/dialect_snowflake.py

```diff
@@ -3840,3 +3887,142 @@ class CallStatementSegment(BaseSegment):
+            ),
+            "FETCH",
+            Indent,
+            OneOf(
+                "FIRST",
+                "NEXT",
+                optional=True,
+            ),
+            Ref("LimitLiteralGrammar"),
+            OneOf(
+                "ROW",
+                "ROWS",
+                optional=True,
+            ),
+            Ref.keyword("ONLY", optional=True),
+            Dedent,
+        ),
+    )
+
+
+@snowflake_dialect.segment(replace=True)
+class SelectClauseSegment(BaseSegment):
+    """A group of elements in a select target statement."""
+
+    type = "select_clause"
+    match_grammar = StartsWith(
+        Sequence("SELECT", Ref("WildcardExpressionSegment", optional=True)),
+        terminator=OneOf(
+            "FROM",
+            "WHERE",
+            Sequence("ORDER", "BY"),
+            "LIMIT",
+            "FETCH",
+            "OFFSET",
+            "OVERLAPS",
+            Ref("SetOperatorSegment"),
+        ),
+        enforce_whitespace_preceding_terminator=True,
+    )
+
+    parse_grammar = Ref("SelectClauseSegmentGrammar")
+
+
+@snowflake_dialect.segment(replace=True)
+class OrderByClauseSegment(BaseSegment):
+    """A `ORDER BY` clause like in `SELECT`."""
+
+    type = "orderby_clause"
+    match_grammar = StartsWith(
+        Sequence("ORDER", "BY"),
+        terminator=OneOf(
+            "LIMIT",
+            "FETCH",
+            "OFFSET",
+            "HAVING",
+            "QUALIFY",
+            # For window functions
+            "WINDOW",
+            Ref("FrameClauseUnitGrammar"),
+            "SEPARATOR",
+        ),
+    )
+    parse_grammar = Sequence(
+        "ORDER",
+        "BY",
+        Indent,
+        Delimited(
+            Sequence(
+                OneOf(
+                    Ref("ColumnReferenceSegment"),
+                    # Can `ORDER BY 1`
+                    Ref("NumericLiteralSegment"),
+                    # Can order by an expression
+                    Ref("ExpressionSegment"),
+                ),
+                OneOf("ASC", "DESC", optional=True),
+                # NB: This isn't really ANSI, and isn't supported in Mysql, but
+                # is supported in enough other dialects for it to make sense here
+                # for now.
+                Sequence("NULLS", OneOf("FIRST", "LAST"), optional=True),
+            ),
+            terminator=OneOf("LIMIT", "FETCH", "OFFSET", Ref("FrameClauseUnitGrammar")),
+        ),
+        Dedent,
+    )
+
+
+@snowflake_dialect.segment(replace=True)
+class HavingClauseSegment(BaseSegment):
+    """A `HAVING` clause like in `SELECT`."""
+
+    type = "having_clause"
+    match_grammar = StartsWith(
+        "HAVING",
+        terminator=OneOf("ORDER", "LIMIT", "FETCH", "OFFSET", "QUALIFY", "WINDOW"),
+        enforce_whitespace_preceding_terminator=True,
+    )
+    parse_grammar = Sequence(
+        "HAVING",
+        Indent,
+        OptionallyBracketed(Ref("ExpressionSegment")),
+        Dedent,
+    )
```

Comment:
These are just to update the terminators.

I did try extracting the terminators as grammar in ANSI (like with `SelectClauseElementTerminatorGrammar`) but that led to strange errors about `Ref` detecting multiple elements so kept as is and copied the segments over

Reply:
If the `parse_grammar` hasn't changed you could copy it, to pick up any future changes to ANSI:

```py
match_grammar = ansi_dialect.get_segment("HavingClauseSegment").match_grammar.copy()
```

But since it's simple enough maybe it's clearer to leave as you have it. WDYT?

Reply:
Sounds good, the only exception is `OrderByClauseSegment` since that also needs the Delimited terminator updated so will leave that one as is 👍